File: library/src/main/java/cn/wandersnail/http/download/DownloadWorker.java
package cn.wandersnail.http.download;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.io.InputStream;
import java.io.RandomAccessFile;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import cn.wandersnail.http.EasyHttp;
import cn.wandersnail.http.TaskInfo;
import cn.wandersnail.http.util.HttpUtils;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

/**
 * date: 2019/8/23 16:17
 * author: zengfansheng
 */
public class DownloadWorker<T extends DownloadInfo> {
    private final Map<T, DownloadController<T>> taskMap = new ConcurrentHashMap<>();
    private final DownloadListener<T> listener;
    private int successNum;
    private int failureNum;
    private boolean isCanceled;
    private final int taskCount;

    public DownloadWorker(@NonNull T info, @Nullable DownloadListener<T> listener) {
        this.listener = listener;
        taskCount = 1;
        info.reset();
        execute(info);
    }

    public DownloadWorker(@NonNull List<T> infoList, @Nullable MultiDownloadListener<T> listener) {
        this.listener = listener;
        taskCount = infoList.size();
        for (T info : infoList) {
            info.reset();
            execute(info);
        }
    }
    
    private void execute(T info) {
        //如果listener为空，说明不需要监听，不为空则在本地监听后，再传出去
        DownloadController<T> controller = new DownloadController<>(info, listener == null ? null : new LocalTaskListener());
        synchronized (this) {
            taskMap.put(info, controller);
        }
        OkHttpClient httpClient = HttpUtils.initHttpsClient(true, new OkHttpClient.Builder())
                .addInterceptor(new ProgressInterceptor(controller))
                .build();
        Call<ResponseBody> call = new Retrofit.Builder()
                .client(httpClient)
                .baseUrl(info.getBaseUrl())
                .build()
                .create(DownloadService.class)
                .download("bytes=" + info.completionLength + "-", info.url);//断点续传
        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(@NonNull Call<ResponseBody> call, @NonNull Response<ResponseBody> response) {
                EasyHttp.executeOnIo(()-> writeToDisk(response.body(), info.getTemporaryFile(), info, controller));
            }

            @Override
            public void onFailure(@NonNull Call<ResponseBody> call, @NonNull Throwable t) {
                controller.onError(t);
            }
        });
    }

    //写入文件在本地
    private void writeToDisk(ResponseBody body, File file, T info, DownloadController<T> controller) {
        if (body == null) {
            controller.onError(new Throwable("body is null"));
            return;
        }
        if (!file.getParentFile().exists()) {
            file.getParentFile().mkdirs();
        }
        long allLen = info.contentLength == 0 ? body.contentLength() : info.contentLength;
        RandomAccessFile accessFile = null;
        FileChannel channel = null;
        InputStream input = null;
        try {
            long readLen = 0;
            input = body.byteStream();
            accessFile = new RandomAccessFile(file, "rwd");
            channel = accessFile.getChannel();
            MappedByteBuffer byteBuffer = channel.map(FileChannel.MapMode.READ_WRITE, info.completionLength, 
                    allLen - info.completionLength);
            byte[] buffer = new byte[10240];
            int len;
            controller.onStart();
            while (!controller.isCancel() && (len = input.read(buffer)) != -1) {
                byteBuffer.put(buffer, 0, len);
                readLen += len;
            }
            if (allLen <= readLen) {
                controller.onComplete();
            }
        } catch (Exception e) {
            Log.e("DownloadWorker", "文件保存失败：" + e.getMessage());
            controller.onError(e);
        } finally {
            HttpUtils.closeQuietly(channel, accessFile, input);
            if (controller.isCancel()) {
                file.delete();
            }
        }
    }
        
    private class LocalTaskListener implements DownloadListener<T> {
        @Override
        public void onStateChange(@NonNull T info, @Nullable Throwable t) {
            if (listener != null) {
                listener.onStateChange(info, t);
            }
            if (info.state == TaskInfo.State.COMPLETED) {
                taskMap.remove(info);
                successNum++;
            } else if (info.state == TaskInfo.State.CANCEL || info.state == TaskInfo.State.ERROR) {
                taskMap.remove(info);
                failureNum++;
            } else if (info.state != TaskInfo.State.START) {
                return;
            }
            if (listener instanceof MultiDownloadListener) {
                ((MultiDownloadListener<T>) listener).onTotalProgress(successNum, failureNum, taskCount);
            }
        }

        @Override
        public void onProgress(@NonNull T info, int progress) {
            if (listener != null) {
                listener.onProgress(info, progress);
            }
        }
    }


    /**
     * 取消单个下载
     */
    public synchronized void cancel(@NonNull T info) {
        DownloadController<T> controller = taskMap.remove(info);
        if (controller != null) {
            controller.cancel();
            if (taskMap.isEmpty()) {
                isCanceled = true;
            }
        }
    }

    /**
     * 取消所有下载
     */
    public synchronized void cancel() {
        if (!taskMap.isEmpty()) {
            for (DownloadController<T> controller : taskMap.values()) {
                controller.cancel();
            }
            taskMap.clear();
            isCanceled = true;
        }
    }

    /**
     * 是否已取消
     */
    public boolean isCanceled() {
        return isCanceled;
    }

    /**
     * 进行中的下载任务数
     */
    public int remaining() {
        return taskMap.size();
    }

    /**
     * 任务是否正在进行
     */
    public boolean isOngoing() {
        return remaining() > 0;
    }

    /**
     * 恢复单个下载
     */
    public synchronized void resume(@NonNull T info) {
        if (info.state == TaskInfo.State.PAUSE) {
            execute(info);
        }
    }

    /**
     * 恢复所有下载
     */
    public synchronized void resume() {
        for (T t : taskMap.keySet()) {
            if (t.state == TaskInfo.State.PAUSE) {
                execute(t);
            }
        }
    }

    /**
     * 暂停单个下载
     */
    public synchronized void pause(@NonNull T info) {
        DownloadController<T> controller = taskMap.get(info);
        if (controller != null) {
            controller.pause();
        }
    }

    /**
     * 暂停所有下载
     */
    public synchronized void pause() {
        for (DownloadController<T> controller : taskMap.values()) {
            controller.pause();
        }
    }
}

File: library/src/main/java/com/snail/network/download/DownloadWorker.kt
package com.snail.network.download

import com.snail.network.TaskInfo
import com.snail.network.exception.RetryWhenException
import com.snail.network.interceptor.ProgressInterceptor
import com.snail.network.utils.HttpUtils
import com.snail.network.utils.IOUtils
import com.snail.network.utils.SchedulerUtils
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.io.File
import java.io.RandomAccessFile
import java.nio.channels.FileChannel
import java.util.concurrent.ConcurrentHashMap

/**
 * 下载任务
 *
 * date: 2019/2/23 18:13
 * author: zengfansheng
 */
class DownloadWorker<T : DownloadInfo> : Disposable {
    private val taskMap = ConcurrentHashMap<T, DownloadObserver<T>>()
    private val listener: DownloadListener<T>?
    private val totalTasks: Int
    private var successCount = 0
    private var failedCount = 0
    private var isCanceled = false
    
    internal constructor(info: T, listener: DownloadListener<T>?) {
        this.listener = listener
        totalTasks = 1
        info.reset()
        this.execute(info)
    }
    
    internal constructor(infos: List<T>, listener: MultiDownloadListener<T>?) {
        this.listener = listener
        totalTasks = infos.size
        infos.forEach {
            it.reset()
            this.execute(it)
        }
    }

    private fun execute(info: T) {
        //如果listener为空，说明不需要监听，不为空则在本地监听后，再传出去
        val observer = DownloadObserver(info, if (listener == null) null else LocalTaskListener())
        taskMap[info] = observer
        val interceptor = ProgressInterceptor(observer)
        val client = HttpUtils.initHttpsClient(true, OkHttpClient.Builder())
            .addInterceptor(interceptor)
            .build()
        Retrofit.Builder()
            .client(client)
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .baseUrl(info.baseUrl)
            .build()
            .create(DownloadService::class.java)
            //断点续传
            .download("bytes=" + info.completionLength + "-", info.url)
            //失败后的retry配置
            .retryWhen(RetryWhenException())
            //写入文件
            .map {
                writeToDisk(it.body(), File(info.temporaryFilePath), info)
                it
            }
            .compose(SchedulerUtils.applyGeneralObservableSchedulers())
            .subscribe(observer)
    }
    
    /**
     * 暂停所有下载
     */
    fun pause() {
        taskMap.values.forEach { 
            it.dispose(false)
        }
    }

    /**
     * 暂停单个下载
     */
    fun pause(info: T) {
        taskMap[info]?.dispose(false)
    }

    /**
     * 恢复所有下载
     */
    fun resume() {
        AndroidSchedulers.mainThread().scheduleDirect {
            taskMap.keys.forEach {
                if (it.state == TaskInfo.State.PAUSE) {
                    execute(it)
                }
            }
        }
    }

    /**
     * 恢复单个下载
     */
    fun resume(info: T) {
        AndroidSchedulers.mainThread().scheduleDirect {
            if (info.state == TaskInfo.State.PAUSE) {
                execute(info)
            }
        }
    }

    private inner class LocalTaskListener : DownloadListener<T> {
        override fun onStateChange(info: T, t: Throwable?) {
            listener?.onStateChange(info, t)
            if (totalTasks > 1) {
                if (info.state == TaskInfo.State.COMPLETED) {
                    taskMap.remove(info)
                    successCount++
                    (listener as? MultiDownloadListener)?.onTotalProgress(successCount, failedCount, totalTasks)
                } else if (info.state == TaskInfo.State.CANCEL || info.state == TaskInfo.State.ERROR) {
                    taskMap.remove(info)
                    failedCount++
                    (listener as? MultiDownloadListener)?.onTotalProgress(successCount, failedCount, totalTasks)
                } else if (info.state == TaskInfo.State.START) {
                    (listener as? MultiDownloadListener)?.onTotalProgress(successCount, failedCount, totalTasks)
                }
            }
        }

        override fun onProgress(info: T) {
            listener?.onProgress(info)
        }
    }

    /**
     * 任务是否正在进行
     */
    fun isOngoing(): Boolean {
        return remaining() > 0
    }

    /**
     * 进行中的下载任务数
     */
    fun remaining(): Int {
        return taskMap.size
    }

    /**
     * 取消所有下载
     */
    @Synchronized
    fun cancel() {
        taskMap.values.forEach { it.dispose(true) }
        taskMap.clear()
        isCanceled = true
    }

    /**
     * 取消单个下载
     */
    @Synchronized
    fun cancel(info: T) {
        taskMap.remove(info)?.dispose(true)
        if (taskMap.isEmpty()) {
            isCanceled = true
        }
    }

    override fun isDisposed(): Boolean {
        return isCanceled
    }

    override fun dispose() {
        cancel()
    }

    //写入文件在本地
    private fun writeToDisk(responseBody: ResponseBody?, file: File, info: T) {
        if (responseBody == null) {
            return
        }
        if (!file.parentFile.exists()) {
            file.parentFile.mkdirs()
        }
        val inputStream = responseBody.byteStream()
        val allLength: Long = if (info.contentLength == 0L) {
            responseBody.contentLength()
        } else {
            info.contentLength
        }
        try {
            RandomAccessFile(file, "rwd").channel.use {
                val mappedBuffer = it.map(FileChannel.MapMode.READ_WRITE, info.completionLength, allLength - info.completionLength)
                val buffer = ByteArray(1024 * 80)
                var len = inputStream.read(buffer)
                while (len != -1) {
                    mappedBuffer.put(buffer, 0, len)
                    len = inputStream.read(buffer)
                }
            }
        } catch (e: Exception) {
            IOUtils.closeQuietly(inputStream)
        }
    }
}